src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardCall.cs

```diff
@@ -0,0 +1,380 @@
+            .WithKindOptions(SymbolDisplayKindOptions.None);
+
+        private static readonly ImmutableHashSet<GuardedCallContext> s_guardedCallContexts = ImmutableHashSet.Create(
+            new GuardedCallContext(
+                DoNotGuardDictionaryRemoveByContainsKeyRule,
+                new ConditionMethodContext(WellKnownTypeNames.SystemCollectionsGenericIDictionary2, "ContainsKey"),
+                ImmutableArray.Create(
+                    new GuardedMethodContext(WellKnownTypeNames.SystemCollectionsGenericIDictionary2, "Remove", 1, ExpectsConditionNegated: false),
+                    new GuardedMethodContext(WellKnownTypeNames.SystemCollectionsGenericDictionary2, "Remove", 2, ExpectsConditionNegated: false))),
+            new GuardedCallContext(
+                DoNotGuardDictionaryRemoveByContainsKeyRule,
+                new ConditionMethodContext(WellKnownTypeNames.SystemCollectionsGenericIReadOnlyDictionary2, "ContainsKey"),
+                ImmutableArray.Create(
+                    new GuardedMethodContext(WellKnownTypeNames.SystemCollectionsImmutableIImmutableDictionary2, "Remove", 1, ExpectsConditionNegated: false))),
+            new GuardedCallContext(
+                DoNotGuardSetAddOrRemoveByContainsRule,
+                new ConditionMethodContext(WellKnownTypeNames.SystemCollectionsGenericICollection1, "Contains"),
+                ImmutableArray.Create(
+                    new GuardedMethodContext(WellKnownTypeNames.SystemCollectionsGenericISet1, "Add", 1, ExpectsConditionNegated: true),
+                    new GuardedMethodContext(WellKnownTypeNames.SystemCollectionsGenericICollection1, "Remove", 1, ExpectsConditionNegated: false))),
+            new GuardedCallContext(
+                DoNotGuardSetAddOrRemoveByContainsRule,
+                new ConditionMethodContext(WellKnownTypeNames.SystemCollectionsImmutableIImmutableSet1, "Contains"),
+                ImmutableArray.Create(
+                    new GuardedMethodContext(WellKnownTypeNames.SystemCollectionsImmutableIImmutableSet1, "Add", 1, ExpectsConditionNegated: true),
+                    new GuardedMethodContext(WellKnownTypeNames.SystemCollectionsImmutableIImmutableSet1, "Remove", 1, ExpectsConditionNegated: false))));
+
+        public sealed override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } =
+            ImmutableArray.Create(DoNotGuardDictionaryRemoveByContainsKeyRule, DoNotGuardSetAddOrRemoveByContainsRule);
+
+        public sealed override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            foreach (var guardedCallContext in s_guardedCallContexts)
```

Comment:
Wonder if you tried/considered to populate all symbols once and register one action to check all, might be more efficient.

Reply:
I was thinking about this in testing as I had a bug where some guarded calls were found twice (two separate actions found them) but did not follow through.
Should I prepare a new PR where everything is consolidated into one action or should we keep it like this?

Reply:
> Should I prepare a new PR where everything is consolidated into one action 

That would be great, thank you!

Reply:
I've tested loading all symbols once and then registering one action like this:

```cs
private void OnCompilationStart(CompilationStartAnalysisContext context)
{
    var guardedCallContextWithSymbols = s_guardedCallContexts
        .Select(c => GuardedCallSymbols.TryGetSymbols(context.Compilation, c, out var symbols)
            ? (Context: c, Symbols: symbols)
            : (Context: c, Symbols: null))
        .WhereAsArray(p => p.Symbols is not null);

    if (!guardedCallContextWithSymbols.IsEmpty)
    {
        context.RegisterOperationAction(OnConditional, OperationKind.Conditional);

        void OnConditional(OperationAnalysisContext context)
        {
            foreach (var (guardedCallContext, symbols) in guardedCallContextWithSymbols)
            {
                var conditional = (IConditionalOperation)context.Operation;

                if (!symbols!.HasApplicableConditionInvocation(conditional.Condition, out var conditionInvocation, out bool containsNegated) ||
                    !symbols.HasApplicableGuardedInvocation(conditional, containsNegated, out var guardedInvocation) ||
                    !AreInvocationsOnSameInstance(conditionInvocation, guardedInvocation) ||
                    !AreInvocationArgumentsEqual(conditionInvocation, guardedInvocation))
                {
                    continue;
                }

                using var locations = ArrayBuilder<Location>.GetInstance(2);
                locations.Add(conditional.Syntax.GetLocation());
                locations.Add(guardedInvocation.Syntax.Parent!.GetLocation());

                context.ReportDiagnostic(conditionInvocation.CreateDiagnostic(
                    guardedCallContext.Rule,
                    additionalLocations: locations.ToImmutable(),
                    properties: null,
                    guardedInvocation.TargetMethod.ToDisplayString(s_symbolDisplayFormat),
                    conditionInvocation.TargetMethod.ToDisplayString(s_symbolDisplayFormat)));
            }
        }
    }
}
```

I've tried to measure the performance difference with `AnalyzerRunner`. The variance between runs of the same code (old one and the updated above) was quite high on my PC (from 5s to 11s for both, tested against `dotnet/roslyn` with a Release build), which makes it kinda hard to say for sure that one is more performant than the other.

Do you know a better way to compare the performance of two analyzers?

Reply:
> Do you know a better way to compare the performance of two analyzers?

I haven't really tried to measure a perf of analyzers, there is some suggestions in https://github.com/dotnet/roslyn-analyzers/blob/main/docs/Performance.md, you might have already checked that. 

We also have some [PerformanceTests](https://github.com/dotnet/roslyn-analyzers/tree/main/src/PerformanceTests) in the repo, by my understanding that runs the [tests](https://github.com/dotnet/roslyn-analyzers/blob/main/src/PerformanceTests/Tests) on CI, saves the results somewhere and compares them with the next PR's perf build and fails if the diff was above the threshold, not sure if adding a test for the analyzer there and use the old version as a `Base` could be a better way. The [Perf.cmd](https://github.com/dotnet/roslyn-analyzers/blob/main/Perf.cmd) script in the root would run all performance tests (which probably would take several hours to finish), unfortunately it seems there is no way to run only one specific perf test  

Ping @mavasani for better suggestions